pybaseballdatana/analysis/simulations/components/state.py

```diff
@@ -0,0 +1,269 @@
+import attr
+from typing import List
+from pybaseballdatana.analysis.simulations.components.event import (
+    BattingEvent,
+    FirstBaseRunningEvent,
+    SecondBaseRunningEvent,
+    ThirdBaseRunningEvent,
+)
+from pybaseballdatana.analysis.simulations.components.player import Batter
+from pybaseballdatana.analysis.utils import check_len, check_is_zero_one
+from functools import partial, lru_cache
+
+
+@attr.s(frozen=True)
+class BaseState:
+    first_base = attr.ib(type=int, validator=check_is_zero_one)
+    second_base = attr.ib(type=int, validator=check_is_zero_one)
+    third_base = attr.ib(type=int, validator=check_is_zero_one)
+
+    def __attrs_post_init__(self):
+        object.__setattr__(
+            self, "bases", (self.first_base, self.second_base, self.third_base)
+        )
+
+
+@lru_cache(maxsize=512)
+def base_out_state_evolve_fun(
+    cls,
+    batting_event,
+    first_base_running_event=FirstBaseRunningEvent.DEFAULT,
+    second_base_running_event=SecondBaseRunningEvent.DEFAULT,
+    third_base_running_event=ThirdBaseRunningEvent.DEFAULT,
+):
+    return cls.evolve(
+        batting_event,
+        first_base_running_event=first_base_running_event,
+        second_base_running_event=second_base_running_event,
+        third_base_running_event=third_base_running_event,
+    )
+
+
+@attr.s(frozen=True)
+class BaseOutState:
+    base_state = attr.ib(type=BaseState)
+    outs = attr.ib(type=int)
+
+    @staticmethod
+    @lru_cache(maxsize=128)
+    def runs_scored(initial_state, end_state):
+        # runs = -d(runners) - d(outs) + 1
+        if end_state.outs == 3:
+            return 0
+        runners_end = sum(end_state.base_state.bases)
```

Comment:
Summing bases can be simplified by adding a `__add__` method to the `BaseState` class

Reply:
correction, you can add a `__iter__` method since I want add to add the int value of the base, not result in a new `BaseState` object